staging/src/k8s.io/apiextensions-apiserver/test/integration/conversion_test.go

```diff
@@ -0,0 +1,281 @@
+	assert.False(t, hostExists)
+	_, portExists := obj.Object["port"]
+	assert.False(t, portExists)
+}
+
+func verifyV2Object(t *testing.T, crd *apiextensionsv1beta1.CustomResourceDefinition, obj *unstructured.Unstructured) {
+	assert.Equal(t, crd.Spec.Group+"/v2", obj.GetAPIVersion())
+	_, hostPortExists := obj.Object["hostPort"]
+	assert.False(t, hostPortExists)
+	host, hostExists := obj.Object["host"]
+	assert.True(t, hostExists)
+	assert.Equal(t, "localhost", host)
+	port, portExists := obj.Object["port"]
+	assert.True(t, portExists)
+	assert.Equal(t, "8080", port)
+}
+
+func testCustomResourceConversionWebhook(t *testing.T, crd *apiextensionsv1beta1.CustomResourceDefinition, customResourceClients map[string]dynamic.ResourceInterface) {
+	name := "cr-instance-1"
+	t.Log("Creating a v1 custom resource")
+	crInstance := &unstructured.Unstructured{
+		Object: map[string]interface{}{
+			"kind":       crd.Spec.Names.Kind,
+			"apiVersion": crd.Spec.Group + "/v1",
+			"metadata": map[string]interface{}{
+				"name": name,
+			},
+			"hostPort": "localhost:8080",
+		},
+	}
+	_, err := customResourceClients["v1"].Create(crInstance, metav1.CreateOptions{})
+	if !assert.NoError(t, err) {
+		t.FailNow()
+	}
+	t.Log("v2 custom resource should be converted")
+	v2cr, err := customResourceClients["v2"].Get(name, metav1.GetOptions{})
+	if !assert.NoError(t, err) {
+		t.FailNow()
+	}
+	if !assert.NotNil(t, v2cr) {
+		t.FailNow()
+	}
+	verifyV2Object(t, crd, v2cr)
+}
+
+func testCRListConversion(t *testing.T, crd *apiextensionsv1beta1.CustomResourceDefinition, apiextensionsClient apiextensionsclientset.Interface, customResourceClients map[string]dynamic.ResourceInterface) {
+	name1 := "cr-instance-1" // is created by testCustomResourceConversionWebhook
+	name2 := "cr-instance-2"
+	t.Log("Creating a v1 custom resource")
+
+	// cr-instance-1 is stored as v1. Now change storage version to v2.
+	if err := retry.RetryOnConflict(retry.DefaultBackoff, func() error {
+		updateCRD, err := apiextensionsClient.ApiextensionsV1beta1().CustomResourceDefinitions().Get(crd.Name, metav1.GetOptions{})
+		if err != nil {
+			t.Fatalf("failed to get CustomResourceDefinition %q: %v", crd.Name, err)
+		}
+
+		// swap storage version
+		updateCRD.Spec.Versions[0].Storage = false
+		updateCRD.Spec.Versions[1].Storage = true
+
+		updateCRD, err = apiextensionsClient.ApiextensionsV1beta1().CustomResourceDefinitions().Update(updateCRD)
+		if err != nil {
+			return err
+		}
+		crd = updateCRD
+		return nil
+	}); err != nil {
+		t.Fatal(err)
+	}
+
+	t.Log("Create a v2 custom resource")
+	v1cr2 := &unstructured.Unstructured{
+		Object: map[string]interface{}{
+			"kind":       crd.Spec.Names.Kind,
+			"apiVersion": crd.Spec.Group + "/v1",
+			"metadata": map[string]interface{}{
+				"name": name2,
+			},
+			"hostPort": "localhost:8080",
+		},
+	}
+
+	// TODO: we have to wait for the storage version to become effective. Storage version changes are not instant.
```

Comment:
if we know we have a race, we should at least add a `time.Sleep(time.Second)` here until we get a more complete way of dealing with it. I'm not eager to add a flake to our integration suite.

Reply:
added a sleep. But why should that flake? I believe it will run through, but will "just" not test what it should.

Reply:
just to give some data point, the old e2e test didn't look flaky from https://testgrid.k8s.io/google-gce#gci-gce-alpha-features, but maybe it's because e2e test is slower